pkg/security/probe/process_killer.go

```diff
@@ -147,3 +191,84 @@ func (p *ProcessKiller) KillAndReport(scope string, signal string, rule *rules.R
 	ev.ActionReports = append(ev.ActionReports, report)
 	p.pendingReports = append(p.pendingReports, report)
 }
+
+// Reset resets the disarmer state
+func (p *ProcessKiller) Reset() {
+	p.ruleDisarmersLock.Lock()
+	clear(p.ruleDisarmers)
+	p.ruleDisarmersLock.Unlock()
+}
+
+type killDisarmer struct {
+	disarmed        bool
+	containerCache  *disarmerCache[string, bool]
+	executableCache *disarmerCache[string, bool]
+}
+
+type disarmerCache[K comparable, V any] struct {
+	*ttlcache.Cache[K, V]
+	capacity uint64
+}
+
+func newDisarmerCache[K comparable, V any](capacity uint64, period time.Duration) *disarmerCache[K, V] {
+	cacheOpts := []ttlcache.Option[K, V]{
+		ttlcache.WithCapacity[K, V](capacity),
+	}
+
+	if period > 0 {
+		cacheOpts = append(cacheOpts, ttlcache.WithTTL[K, V](period))
+	}
+
+	return &disarmerCache[K, V]{
+		Cache:    ttlcache.New[K, V](cacheOpts...),
+		capacity: capacity,
+	}
+}
+
+func newKillDisarmer(cfg *config.RuntimeSecurityConfig) *killDisarmer {
+	kd := &killDisarmer{
+		disarmed: false,
+	}
+
+	if cfg.EnforcementDisarmerContainerEnabled {
+		capacity := uint64(cfg.EnforcementDisarmerContainerMaxAllowed)
+		if capacity == 0 {
+			capacity = defaultDisarmerContainerMaxAllowed
+		}
+		kd.containerCache = newDisarmerCache[string, bool](capacity, cfg.EnforcementDisarmerContainerPeriod)
+	}
+
+	if cfg.EnforcementDisarmerExecutableEnabled {
+		capacity := uint64(cfg.EnforcementDisarmerExecutableMaxAllowed)
+		if capacity == 0 {
+			capacity = defaultDisarmerExecutableMaxAllowed
+		}
+		kd.executableCache = newDisarmerCache[string, bool](capacity, cfg.EnforcementDisarmerExecutablePeriod)
+	}
+
+	return kd
+}
+
+func (kd *killDisarmer) allow(cache *disarmerCache[string, bool], key string, onDisarm func()) bool {
+	if kd.disarmed {
+		return false
+	}
+
+	if cache == nil {
+		return true
+	}
+
+	cache.DeleteExpired()
+	// if the key is not in the cache, check if the new key causes the number of keys to exceed the capacity
+	// otherwise, the key is already in the cache and cache.Get will update its TTL
+	if cache.Get(key) == nil {
+		alreadyAtCapacity := uint64(cache.Len()) >= cache.capacity
+		cache.Set(key, true, ttlcache.DefaultTTL)
+		if alreadyAtCapacity && !kd.disarmed {
+			kd.disarmed = true
```

Comment:
I would add a way to re-arm, for example where there is a ruleset_loaded or when all the key expired

Reply:
The rules are already being re-armed when a new ruleset is loaded through the `Reset()` function [here](https://github.com/DataDog/datadog-agent/pull/28970/files#diff-2fd6ed204a0c7356424c077b41af3af49319295624726c218b6a89454675bd37R195-R200).
I added a mechanism to re-arm rules once all keys have expired with commit fb48cd83bca32033b7069f1490f17e67cafe94e9.